src/sensors/IMD.py

```diff
@@ -29,19 +34,16 @@ def __init__(
 
     def _IMDInput(self, frequency, duty_cycle, good_state: bool) -> Input:
         if self.power_on.get_pin_state() == memory.DigitalOut.State.Low:
-            raise InputFailedException
-        elif good_state is True:
-            return Multiple(
-                self.ok_hs.generate_state(memory.DigitalIn.State.High),
-                self.m_hs.generate_duty(duty_cycle),
-                self.m_hs.generate_frequency(frequency),
-            )
-        else:
-            return Multiple(
-                self.ok_hs.generate_state(memory.DigitalIn.State.Low),
-                self.m_hs.generate_duty(duty_cycle),
-                self.m_hs.generate_frequency(frequency),
-            )
+            raise IMDException
+        return Multiple(
+            self.ok_hs.generate_state(
+                memory.DigitalIn.State.High
+                if good_state
+                else memory.DigitalIn.State.Low
+            ),
+            self.m_hs.generate_duty(duty_cycle),
+            self.m_hs.generate_frequency(frequency),
+        )
```

Comment:
hmm, now that I look at it, the exception should be thrown when calling the input apply method, we might have to add a checked input condition that checks some condition before the apply. It would be to add a new input that takes an input and a function that will be called before applying the input, the function itself will throw the exception if needed. I can do this, but seeing that you need it I think it wouldn't take much time to do it yourself, but I can do it myself this afternoon